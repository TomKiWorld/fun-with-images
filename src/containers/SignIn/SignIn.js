import React, { Component } from 'react';
import FormInput from '../../components/FormFields/FormInput/FormInput';
import FormSubmit from '../../components//FormFields/FormSubmit/FormSubmit';
import Preloader from '../../components/Preloader/Preloader';
import { DATABASE } from '../../constants';

/**
 * Sign In component
 * - Allows a user to sign in to the app by displaying a form
 * 
 * Required props:
 * - loadUser => Functon to get user information from state
 * - onRouteChange => Function to change the state of route
 */
class SignIn extends Component {
  constructor() {
    super();
    this.state = {
      signInEmail: '',
      signInPassword: '',
      loginError: '',
      loading: false
    }
  }

  // Set state on input change
  onFieldChange = (event, term) => {
    this.setState({[term]: event.target.value })
  }

  // Sign in 
  onSubmitSignIn = (e, email, password) => {
    e.preventDefault();
    if ((email !== '') && (password !== '')) {
      this.setState({loading: true})
      fetch(`${DATABASE}/signin`, {
        method: 'post',
        headers: {'Content-type': 'application/json'},
        body: JSON.stringify({
          email: email,
          password: password
        })
      })
      .then(response => response.json())
      .then(user => {
        if (user.id) {
          this.props.loadUser(user);
          this.props.onRouteChange('home');
        } else {
          this.setState({loading: false})
          this.setState({loginError: user});
        }
      })
    } else {
      this.setState({loginError: 'Please make sure both fields are entered'});
    }
  }

  render() {
    const { onRouteChange } = this.props;
    const loginError = this.state.loginError ? <p className='error-message'>{this.state.loginError}</p> : '';
    const preload = this.state.loading ? <Preloader/> : '';

    return (
      <article className='sign-in br3 ba dark-gray b--black-10 mv4 mw6 shadow-5 center'>
        <form className='pa4 black-80' action='/'>
          <div className='measure'>
            <fieldset id='sign_up' className='ba b--transparent ph0 mh0'>
              <h1 className='f1 fw6 ph0 mh0'>Sign In</h1>
              <FormInput 
                label='Email'
                name='email-address'
                type='email'
                onChange={(e) => this.onFieldChange(e, 'signInEmail')}
                extraClass='inputField'
              />
              <FormInput 
                label='Password'
                name='password'
                type='password'
                onChange={(e) => this.onFieldChange(e, 'signInPassword')}
                extraClass='inputField'
              />
            </fieldset>
            {loginError}
            <FormSubmit
              value='Sign in'
              type='submit'
              onClick={(e) => this.onSubmitSignIn(e, this.state.signInEmail, this.state.signInPassword)}
              extraClass='signIn'
            />
            <FormSubmit
              value='Log in as Visitor'
              type='submit'
              onClick={(e) => this.onSubmitSignIn(e, 'visitor@gmail.com', 'visit')}
            />
            <div className='lh-copy mt3'>
              <p 
                onClick={() => onRouteChange('register')}
                className='f6 link dim black db pointer'>Register</p>
            </div>
          </div>
        </form>
        {preload}
      </article>
    );
  }  
}

export default SignIn;
