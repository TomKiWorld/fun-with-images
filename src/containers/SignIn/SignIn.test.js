import { shallow } from 'enzyme';
import React from 'react';
import SignIn from './SignIn';

describe('SignIn Component', () => {
  const wrapper = shallow(<SignIn />);
  const nameInput = wrapper.find('[name=\'email-address\']').at(0);
  const passwordInput = wrapper.find('[name=\'password\']').at(0);

  it('Expect to render SignIn Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Expect to handle error submit sign in', () => {
    const submitBtn = wrapper.find('[value=\'Sign in\']').at(0);
    submitBtn.simulate('click', { preventDefault: jest.fn() });
    expect(wrapper.state().loginError).toEqual('Please make sure both fields are entered');
  });

  it('Expect to handle email change', () => {
    nameInput.simulate('change', { target: { value: 'john@gmail.com', term: 'email'} });
    expect(wrapper.state().signInEmail).toEqual('john@gmail.com');
  });

  it('Expect to handle password change', () => {
    passwordInput.simulate('change', { target: { value: 'unsafePassword', term: 'password'} });
    expect(wrapper.state().signInPassword).toEqual('unsafePassword');
  });

  it('Expect to handle submit sign in', () => {
    const submitBtn = wrapper.find('[value=\'Sign in\']').at(0);
    submitBtn.simulate('click', { preventDefault: jest.fn() });
    expect(wrapper.state().loading).toEqual(true);
  });

  it('Expect to handle submit sign in as visitor', () => {
    const submitBtn = wrapper.find('[value=\'Log in as Visitor\']').at(0);
    submitBtn.simulate('click', { preventDefault: jest.fn() });
    expect(wrapper.state().loading).toEqual(true);
  });
});
