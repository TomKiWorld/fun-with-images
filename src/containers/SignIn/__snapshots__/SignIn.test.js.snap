// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Component Expect to render SignIn Component 1`] = `ShallowWrapper {}`;
